File: trunk/java/pdfclown.lib/src/org/pdfclown/documents/contents/objects/SetLineDash.java
/*
  Copyright 2007-2011 Stefano Chizzolini. http://www.pdfclown.org

  Contributors:
    * Stefano Chizzolini (original code developer, http://www.stefanochizzolini.it)

  This file should be part of the source code distribution of "PDF Clown library"
  (the Program): see the accompanying README files for more info.

  This Program is free software; you can redistribute it and/or modify it under the terms
  of the GNU Lesser General Public License as published by the Free Software Foundation;
  either version 3 of the License, or (at your option) any later version.

  This Program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY,
  either expressed or implied; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.

  You should have received a copy of the GNU Lesser General Public License along with this
  Program (see README files); if not, go to the GNU website (http://www.gnu.org/licenses/).

  Redistribution and use, with or without modification, are permitted provided that such
  redistributions retain the above copyright notice, license and disclaimer, along with
  this list of conditions.
*/

package org.pdfclown.documents.contents.objects;

import java.util.List;

import org.pdfclown.PDF;
import org.pdfclown.VersionEnum;
import org.pdfclown.documents.contents.ContentScanner.GraphicsState;
import org.pdfclown.documents.contents.LineDash;
import org.pdfclown.objects.PdfArray;
import org.pdfclown.objects.PdfDirectObject;
import org.pdfclown.objects.PdfNumber;
import org.pdfclown.objects.PdfReal;

/**
  'Set the line dash pattern' operation [PDF:1.6:4.3.3].

  @author Stefano Chizzolini (http://www.stefanochizzolini.it)
  @since 0.0.4
  @version 0.1.1, 03/22/11
*/
@PDF(VersionEnum.PDF10)
public final class SetLineDash
  extends Operation
{
  // <class>
  // <static>
  // <fields>
  public static final String Operator = "d";
  // </fields>
  // </static>

  // <dynamic>
  // <constructors>
  public SetLineDash(
    float phase,
    float unitsOn,
    float unitsOff
    )
  {
    super(
      Operator,
      new PdfArray(
        new PdfReal(unitsOn),
        new PdfReal(unitsOff)
        ),
      new PdfReal(phase)
      );
  }

  public SetLineDash(
    List<PdfDirectObject> operands
    )
  {super(Operator, operands);}
  // </constructors>

  // <interface>
  // <public>
  @Override
  public void scan(
    GraphicsState state
    )
  {state.setLineDash(getValue());}

  public LineDash getValue(
    )
  {
    // 1. Dash array.
    PdfArray baseDashArray = (PdfArray)operands.get(0);
    float[] dashArray = new float[baseDashArray.size()];
    for(
      int index = 0,
        length = dashArray.length;
      index < length;
      index++
      )
    {dashArray[index] = ((PdfNumber<?>)baseDashArray.get(index)).getNumberValue();}
    // 2. Dash phase.
    float dashPhase = ((PdfNumber<?>)operands.get(1)).getNumberValue();

    return new LineDash(dashArray, dashPhase);
  }

  public void setValue(
    LineDash value
    )
  {
    // 1. Dash array.
    float[] dashArray = value.getDashArray();
    PdfArray baseDashArray = new PdfArray(dashArray.length);
    for(
      int index = 0,
        length = dashArray.length;
      index < length;
      index++
      )
    {baseDashArray.set(index, new PdfReal(dashArray[index]));}
    operands.set(0,baseDashArray);
    // 2. Dash phase.
    operands.set(1, new PdfReal(value.getDashPhase()));
  }
  // </public>
  // </interface>
  // </dynamic>
  // </class>
}

File: trunk/dotNET/pdfclown.lib/src/org/pdfclown/documents/contents/objects/SetLineDash.cs
/*
  Copyright 2007-2012 Stefano Chizzolini. http://www.pdfclown.org

  Contributors:
    * Stefano Chizzolini (original code developer, http://www.stefanochizzolini.it)

  This file should be part of the source code distribution of "PDF Clown library" (the
  Program): see the accompanying README files for more info.

  This Program is free software; you can redistribute it and/or modify it under the terms
  of the GNU Lesser General Public License as published by the Free Software Foundation;
  either version 3 of the License, or (at your option) any later version.

  This Program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY,
  either expressed or implied; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.

  You should have received a copy of the GNU Lesser General Public License along with this
  Program (see README files); if not, go to the GNU website (http://www.gnu.org/licenses/).

  Redistribution and use, with or without modification, are permitted provided that such
  redistributions retain the above copyright notice, license and disclaimer, along with
  this list of conditions.
*/

using org.pdfclown.bytes;
using org.pdfclown.objects;

using System.Collections.Generic;

namespace org.pdfclown.documents.contents.objects
{
  /**
    <summary>'Set the line dash pattern' operation [PDF:1.6:4.3.3].</summary>
  */
  [PDF(VersionEnum.PDF10)]
  public sealed class SetLineDash
    : Operation
  {
    #region static
    #region fields
    public static readonly string OperatorKeyword = "d";
    #endregion
    #endregion

    #region dynamic
    #region constructors
    public SetLineDash(
      double phase,
      double unitsOn,
      double unitsOff
      ) : base(
        OperatorKeyword,
        new PdfArray(
          PdfReal.Get(unitsOn),
          PdfReal.Get(unitsOff)
          ),
        PdfReal.Get(phase)
        )
    {}

    public SetLineDash(
      IList<PdfDirectObject> operands
      ) : base(OperatorKeyword, operands)
    {}
    #endregion

    #region interface
    #region public
    public override void Scan(
      ContentScanner.GraphicsState state
      )
    {state.LineDash = Value;}

    public LineDash Value
    {
      get
      {
        // 1. Dash array.
        PdfArray baseDashArray = (PdfArray)operands[0];
        double[] dashArray = new double[baseDashArray.Count];
        for(
          int index = 0,
            length = dashArray.Length;
          index < length;
          index++
          )
        {dashArray[index] = ((IPdfNumber)baseDashArray[index]).RawValue;}
        // 2. Dash phase.
        double dashPhase = ((IPdfNumber)operands[1]).RawValue;

        return new LineDash(dashArray, dashPhase);
      }
      set
      {
        // 1. Dash array.
        double[] dashArray = value.DashArray;
        PdfArray baseDashArray = new PdfArray(dashArray.Length);
        for(
          int index = 0,
            length = dashArray.Length;
          index < length;
          index++
          )
        {baseDashArray[index] = PdfReal.Get(dashArray[index]);}
        operands[0] = baseDashArray;
        // 2. Dash phase.
        operands[1] = PdfReal.Get(value.DashPhase);
      }
    }
    #endregion
    #endregion
    #endregion
  }
}